cpp/include/cudf/copying.hpp

```diff
@@ -267,6 +267,64 @@ std::vector<column_view> split(column_view const& input,
+ *          @p rhs based on the value of the corresponding element in @p boolean_mask
+ *
+ * Selects each element i in the output column from either @p rhs or @p lhs using the following rule:
+ *          output[i] = (boolean_mask[i]) ? lhs : rhs[i]
+ *         
+ * @throws cudf::logic_error if lhs and rhs are not of the same type 
+ * @throws cudf::logic_error if boolean mask is not of type bool8
+ * @throws cudf::logic_error if boolean mask is not of the same length as rhs  
+ * @param[in] left-hand scalar
+ * @param[in] right-hand column_view
+ * @param[in] column_view representing "left (true) / right (false)" boolean for each element
+ * @param[in] mr resource for allocating device memory 
+ *
+ * @returns new column with the selected elements
+ */
+std::unique_ptr<column> copy_if_else(scalar const& lhs, column_view const& rhs, column_view const& boolean_mask,
+                                    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief   Returns a new column, where each element is selected from either @p lhs or 
+ *          @p rhs based on the value of the corresponding element in @p boolean_mask
+ *
+ * Selects each element i in the output column from either @p rhs or @p lhs using the following rule:
+ *          output[i] = (boolean_mask[i]) ? lhs[i] : rhs
+ *         
+ * @throws cudf::logic_error if lhs and rhs are not of the same type 
+ * @throws cudf::logic_error if boolean mask is not of type bool8
+ * @throws cudf::logic_error if boolean mask is not of the same length as lhs  
+ * @param[in] left-hand column_view
+ * @param[in] right-hand scalar
+ * @param[in] column_view representing "left (true) / right (false)" boolean for each element
+ * @param[in] mr resource for allocating device memory 
+ *
+ * @returns new column with the selected elements
+ */
+std::unique_ptr<column> copy_if_else(column_view const& lhs, scalar const& rhs, column_view const& boolean_mask,
+                                    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource());
+                                    
+/**
+ * @brief   Returns a new column, where each element is selected from either @p lhs or 
+ *          @p rhs based on the value of the corresponding element in @p boolean_mask
+ *
+ * Selects each element i in the output column from either @p rhs or @p lhs using the following rule:
+ *          output[i] = (boolean_mask[i]) ? lhs : rhs
+ *          
+ * @throws cudf::logic_error if boolean mask is not of type bool8 
+ * @param[in] left-hand scalar
+ * @param[in] right-hand scalar
+ * @param[in] column_view representing "left (true) / right (false)" boolean for each element
+ * @param[in] mr resource for allocating device memory 
+ *
+ * @returns new column with the selected elements
+ */
+std::unique_ptr<column> copy_if_else( scalar const& lhs, scalar const& rhs, column_view const& boolean_mask,
```

Comment:
Is this one necessary? It's really just a gather. Target column has `N` elements, Source column has `2` elements. gather map has `N` elements, with value 0 or 1 (same as a boolean mask).

Reply:
Adding jlowe.
@jlowe

Reply:
This isn't technically necessary but makes it a bit easier.  From the Spark plugin perspective, three expressions lead up to this call.  The first two expressions will either evaluate to a column or a scalar, while the third expression will always be a predicate expression (i.e.: a boolean column).

We can definitely special-case when the two expressions are both scalars to handle this.  The process, as I understand it, would be:
- Allocate a temporary two-column device vector with a validity buffer
- Perform two tiny device memcpy's to construct the two-element column data vector
- Build the validity mask for the two-element column (assume on the host and sent down to device)
- Cast the boolean column to an integer column ([gather fails if the map is boolean](https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/src/copying/gather.cu#L85))
- Call gather with the two-element column and mask-as-integer column
- Free the temporary two-element and cast result columns

I was hoping that implementing the scalar-scalar version of this would be trivial once column-scalar was implemented.  If that's not the case, we can use the gather workaround instead.

Reply:
I think it's fine to include it. But I guess you have a choice of how you implement it internally).